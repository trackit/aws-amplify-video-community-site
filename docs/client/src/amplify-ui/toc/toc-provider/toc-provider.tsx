import {Component, State, Watch, h, Build} from "@stencil/core";
import {tocContext} from "../toc.context";
import {SetContent} from "../toc.types";
import {headingIsVisible} from "../../../utils/heading-is-visible";

@Component({tag: "amplify-toc-provider", shadow: false})
export class AmplifyTOCProvider {
  /** The slot within `amplify-toc-content` */
  @State() content?: HTMLElement;
  /** Extracted, visible h2 & h3s within `this.content` */
  @State() elements?: HTMLElement[];

  observer?: MutationObserver;

  setContent: SetContent = (content: HTMLElement) => (this.content = content);

  setElements = () => {
    if (this.content) {
      this.elements = Array.from(
        this.content.querySelectorAll("h2, h3"),
      ).filter((e) => headingIsVisible(e)) as HTMLElement[];
    }
  };

  @Watch("content")
  bindToContent() {
    this.setElements();
    if (this.observer) this.observer.disconnect();
    if (Build.isBrowser) {
      this.observer = new MutationObserver(this.setElements);
      this.content &&
        this.observer.observe(this.content, {
          childList: true,
          attributes: true,
          subtree: true,
        });
    }
  }

  componentDidUnload() {
    this.observer?.disconnect();
  }

  render() {
    return (
      <tocContext.Provider
        state={{
          elements: this.elements,
          setContent: this.setContent,
        }}
      >
        <slot />
      </tocContext.Provider>
    );
  }
}
