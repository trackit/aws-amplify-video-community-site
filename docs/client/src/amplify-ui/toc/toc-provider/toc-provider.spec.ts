import {AmplifyTOCProvider} from "./toc-provider";
import {newSpecPage} from "@stencil/core/testing";

describe("amplify-toc-provider", () => {
  describe("Component logic", () => {
    let tocProvider: AmplifyTOCProvider;
    beforeEach(() => (tocProvider = new AmplifyTOCProvider()));

    it("should init `content` as `undefined`", () =>
      expect(tocProvider.content).toBeUndefined());

    it("should init `elements` as `undefined`", () =>
      expect(tocProvider.elements).toBeUndefined());
  });

  describe("Render logic", () => {
    it("should render", async () => {
      expect(
        (
          await newSpecPage({
            components: [AmplifyTOCProvider],
            html: `<amplify-toc-provider />`,
          })
        ).root,
      ).toMatchSnapshot();
    });
  });
});
