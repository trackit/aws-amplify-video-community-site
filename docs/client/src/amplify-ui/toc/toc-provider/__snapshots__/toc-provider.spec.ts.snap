// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amplify-toc-provider Render logic should render 1`] = `
<amplify-toc-provider>
  <!---->
</amplify-toc-provider>
`;
